### ISSUE TITLE
Restarting after aligning reads to the graph

### ISSUE THREAD

**Issue by shanesturrock** [fenderglass/ABruijn#267]
We're needing to restart the machine I've been running FLYE 2.6 on for a few months and it just finishing up the "Aligning reads to the graph" step. I'm also aware there's a bug with large repetitive genomes which is fixed in 2.7.1 so would I be able to stop and resume using the newer version to avoid the problem with the slow writing of the gfa file?

**Comment by mikolmogorov** [comment]
Unfortunately, the closest possible restart point would be the beginning of repeat graph analysis (`--resume-from repeat`). I definitely recommend to use the version 2.7.1 for that.

What was the bottleneck of your run that terminated? If you can post the log file, I might have some extra suggestions.

Best,
Mikhail

**Comment by shanesturrock** [comment]
Hi Mikhail, here's the log as it currently stands. FLYE 2.6 is still running but the step from 80% to 90% took three weeks to give you an idea of the performance we're seeing. The machine it is on has 72 cores (144 threads) and 6TB of RAM. The repeat stage has been running since December so it is a bit of a blow if we have to go back to then unless the 2.7.1 version moves much faster. The 20-repeat directory has no files in it as yet.

Here's stdout:

[2019-12-22 00:01:29] INFO: Computing consensus
[2019-12-24 16:45:07] INFO: Alignment error rate: 0.104049
[2019-12-24 17:29:36] INFO: >>>STAGE: repeat
[2019-12-24 17:29:36] INFO: Building and resolving repeat graph
[2019-12-24 17:29:42] INFO: Reading sequences
[2019-12-24 18:52:54] INFO: Building repeat graph
0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100%
[2020-02-09 19:25:19] INFO: Median overlap divergence: 0.0609317
[2020-02-12 09:41:37] INFO: Aligning reads to the graph
0% 10% 20% 30% 40% 50% 60% 70% 80% 90%

flye.log has this but hasn't been written to since February:

[2020-02-12 09:30:05] DEBUG: Total edges: 1393150
[2020-02-12 09:35:12] DEBUG: Building positional index
[2020-02-12 09:35:16] DEBUG: Total sequence: 25706452598 bp
[2020-02-12 09:41:37] INFO: Aligning reads to the graph
[2020-02-12 09:43:16] DEBUG: Hard threshold set to 1
[2020-02-12 09:43:16] DEBUG: Started k-mer counting
[2020-02-12 10:05:47] DEBUG: Repetitive k-mer frequency: 778
[2020-02-12 10:05:47] DEBUG: Filtered 1716743 repetitive k-mers (0.00105381)
[2020-02-12 10:06:07] DEBUG: Sampling rate: 2
[2020-02-12 10:06:07] DEBUG: Solid k-mers: 1627358628
[2020-02-12 10:06:07] DEBUG: K-mer index size: 8481063719
[2020-02-12 10:06:07] DEBUG: Mean k-mer frequency: 5.21155
[2020-02-12 10:37:50] DEBUG: Sorting k-mer index

**Comment by mikolmogorov** [comment]
I see. How big is the genome and what type of reads are you using?

Version 2.7 will eliminate on of the bottlenecks which you'd most likely faced in your run, but the compute time for read alignment should be roughly the same, unfortunately.

One thing to check is how many threads to use. We have realy used more than 50 on our hardware. It could be that at some point performance might stop scaling with threads increase and can actually go down. I'd check that Flye is utilizing all CPUs that you have requested at 100%. If not, it's possible that reducing the number of threads might actually help to make the assembly faster. Also, if you see that processes are spending a lot of time in system space, rather than user space, this could also be an indicator that the number of threads needs to be reduced. Finally, extra threads use more memory - check that you are not hitting the system limits. If there are other memory intensive software running on the same machine, it might also slow Flye down significantly because of memory fragmentation.

Best,
Mikhail

**Comment by shanesturrock** [comment]
Hi Mikhail, yeah, the genome is 25Gbp give or take. I had previously tried FLYE and you had baulked at the size and said it wouldn't work but that was a while back. I had been using MaSuRCA as the assembler but while that was running Aleksey had swapped out CABOG for Flye as the [assembler](http://masurca.blogspot.com/2019/03/masurca-version-331.html) and provided instructions on using Flye with the intermediate file from MaSuRCA which is what I did. Flye is certainly using all cores (I allocated 64 cores and I am seeing close to 6400% utilisation and 700GB of RAM usage (the machine has 6TB so not even close). Here's the command that Aleksey recommended using to run Flye with the error corrected data:

GS=`cat ESTIMATED_GENOME_SIZE.txt` && <flye_path>/bin/flye --nano-corr <mr.41.15.15.0.02.1.fa or mr.41.15.17.0.029.1.fa> -t <number_of-threads> -g $GS -m 2000 -o flye_assembly -i 0

If we're to restart the assembly with Flye it would also be helpful if you could comment on what Aleksey is recommending here in case there are further optimisations that will help the job complete in a reasonable time and with a good assembly.

**Comment by mikolmogorov** [comment]
Thanks. I actually do have some good news for the corrected reads assembly - I've been working on integrating minimizer indexing for this kind of input, which should improve speed and memory usage substantially. It's not quite ready yet, but in a couple of weeks we should have a first beta version.

In the mean time, if you have the resources, I'd suggest to run the latest release anyway. The command line looks good, but I suggest to use Flye 2.7.1 (`--resume-from repeat`). Since the pipeline has changed a bit, you'd need to override its version in the `params.json` file (change pipeline version from 2 to 3).

**Comment by shanesturrock** [comment]
Hi Mikhail, thanks, the update sounds worth waiting for. We have to close the machine down in the near term so restarting will have to wait and hopefully your new release will be out by then. If it is still in beta when we move I'll be happy to try it and give feedback.

**Comment by shanesturrock** [comment]
We've now completed our system move and updated all machines so we're ready to start running our assembly again. At the moment 2.7.1 still appears to be the latest but I would be keen to try your beta of the new release if it is ready to go?

**Comment by mikolmogorov** [comment]
@shanesturrock I would try the latest code from `flye` branch, as it contains many relevant updates. The code should be relatively stable, and I expect to make a release in the near future. You can use `--restart-from repeat` option. Let me know how it goes.

**Comment by shanesturrock** [comment]
Thanks Mikhail, I'll get that going today.

**Comment by shanesturrock** [comment]
Hi Mikhail, I've restarted having built from source and changed the json file to pipeline 3. I also allocated 128 cores (double my last run) and so far it is moving along at a much better pace. The "Building repeat graph" step took 6 weeks on 64 cores last time, this time it took 2 days so that's a substantial optimisation. The next step computing gluepoints took three days but that only used a single core and I'm expecting it to take about the same time. Aligning reads to the graph took four months so I'm hoping you've made similar optimisations to this step too. The code got stuck at the resolving repeats last time as that's what it was doing, and had been for around a month when we shut the machine down.

**Comment by mikolmogorov** [comment]
@shanesturrock thanks for the update - keep us posted!

**Comment by shanesturrock** [comment]
One week in and it has progressed through 90% of aligning reads to the graph. This step started two days ago and looks likely to finish in the next day. With 2.6 it took four months on 64 cores. What I am noticing is that the number of repetitive K-mers filtered is substantially higher 2.6 - Filtered 1716743 repetitive k-mers (0.00105381), this version - Filtered 2008094843 repetitive k-mers (0.288154). That's undoubtably contributed greatly to the increase in speed but it will be interesting to see how that affects the number of reads aligned which I should know in a day or so.

**Comment by shanesturrock** [comment]
It's now finished the aligning reads to the graph step which was very quick. I noticed that only about half as many reads aligned versus the last run but this may be intended as it certainly improves performance.

Stats for the old run: [2020-06-06 10:46:41] INFO: Aligned read sequence: 269105407044 / 277711296371 (0.969011)
and for the new run: [2020-07-20 14:24:12] INFO: Aligned read sequence: 146639839881 / 277711296371 (0.52803)

The resulting dot file from 2.6 is 14M in size and the new one is 1.7M. Not sure how this will affect the final assembly but at this point I consider an assembly better than no assembly and we could probably tweak the settings and try again once this completes. The current step shows Simplifying the graph and has been running for nearly a day on a single core.

**Comment by mikolmogorov** [comment]
@shanesturrock thanks for the update!

Happy to hear that it ran much faster this time, but the amount of unaligned sequence is certainly unintended.. I haven't seen this on our datasets yet (but they are not super repetitive).

 Wrt to k-mer filtering - I did change the filtering method slightly, but I also changed the way it's reported. Previously, Flye output the number of unique filtered k-mers, and now it outputs the number of filtered k-mer copies (or genomic positions). Since a few highly repetitive k-mers can contribute to many more positions in the genome, the numbers are inflated.

Do you mind sending me the log file of the current run? If you happen to have the logs from the previous runs - this will be very helpful too.

Best,
Mikhail

**Comment by shanesturrock** [comment]
My data set is around 80% repeats so has always been a challenge for assembly. Here are the log files.

[flye_2.7.1+_run_restart.log](https://github.com/fenderglass/Flye/files/4951165/flye_2.7.1%2B_run_restart.log)
[flye-2.6_run.log](https://github.com/fenderglass/Flye/files/4951167/flye-2.6_run.log)

**Comment by mikolmogorov** [comment]
@shanesturrock thanks, but this is just terminal output? Could you send flye.log files from the output directories?

**Comment by shanesturrock** [comment]
They're 274MB compressed into a zip so I can't attach them here. See if this works: https://we.tl/t-0RfLfVe9ID

**Comment by mikolmogorov** [comment]
@shanesturrock thank you!

It seems that the problem is not in repetitive k-mers being filtered, but rather in error rate thresholds. `-pacbio-corr` mode assumes 1-3% errors left in reads, but based on the distributions from the logs the error rate varies from 1% to 10% (3.3% mean). Previous versions of Flye did not implement error rate threshold cutoffs for aligning reads to the graph, hence the difference in the amount of aligned sequence.

MaSuRCA typically produces high quality reads, but I can see that it might be more challenging for highly repetitive genomes (less unique k-mers for initial mapping), and thus the quality of reads deteriorates. Ideally, given with these reads `--pacbio-raw` mode should be used. It makes sense to wait until this run finishes though. So far, seems like the graph operations are being quite slow - let's see if they will complete within a reasonable time...

**Comment by shanesturrock** [comment]
OK, that makes sense. I've done some error correction using FMLRC using our Illumina data against the PacBio reads and that doesn't seem to have error corrected all that well either. The simplifying graph step is toiling but I'm willing to let it continue. If I run with --pacbio-raw I assume I'll need to start completely again? Also, a more sensitive mode is going to take a performance hit I expect? We're under pressure to get this finished but the data just keeps fighting every step of the way.

**Comment by mikolmogorov** [comment]
If this is going to be the first assembly, it definitely makes sense to wait until the current run finishes. It might not be optimal in terms of contiguity, but I won't be "wrong" (e.g. contain major errors). Once you have a first iteration of the assembly, you can perhaps do another one with more optimized parameters.

Wrt to the performance - it may or may not be slower with `--pacbio-raw`, as the new code includes some optimizations for disjointig assembly step as well. 

Wrt to the current run - feel free to send me the updated log again if it'll be taking too long - will be happy to take a look on what's going on.

**Comment by shanesturrock** [comment]
This will be the first completed with the PacBio data. We did an earlier assembly with just the Illumina reads using SOAPdenovo2 (the only program that could do it and not blow up due to the volume of reads) and that was highly fragmented (around 25 million contigs) and missing about 25% of the the expected sequence. The RAM usage of the current simplifying the graph step is climbing up past 2.3TB and still going. We have 6TB though but I'm hoping it will finish this step within our resources and move on.

**Comment by shanesturrock** [comment]
One week running Simplifying the graph step and it has hit 50% of our RAM (3 out of 6TB). Is there any way to estimate progress and when it is likely to finish? It is only using a single CPU but the memory usage is a concern.

**Comment by mikolmogorov** [comment]
@shanesturrock definitely unexpected performance - could you send me the latest log?

**Comment by shanesturrock** [comment]
Last thing written to the log was on the 20th so I believe you've got everything already since I sent the logs on the 21st. Here's the tail end of the log:

[2020-07-20 20:18:03] DEBUG: Mult: 473524	115998	17	 (1,0)
[2020-07-20 20:18:03] DEBUG: Starting -1822803 aln:21 minSpan:0
[2020-07-20 20:18:03] DEBUG: 	      -764532	num:14	flank:11895	span:8
[2020-07-20 20:18:03] DEBUG: 	      -764531	num:6	flank:2991	span:88
[2020-07-20 20:18:03] DEBUG: Mult: 1822803	144525	17	 (1,0)
[2020-07-20 20:18:03] DEBUG: Starting 1791243 aln:20 minSpan:0
[2020-07-20 20:18:03] DEBUG: 	      -1645018	num:7	flank:16455	span:9
[2020-07-20 20:18:03] DEBUG: 	      -1645017	num:8	flank:9304	span:9
[2020-07-20 20:18:03] DEBUG: Mult: 1791243	148811	12	 (0,1)
[2020-07-20 20:20:01] DEBUG: Writing Dot

and here's the tail end of the terminal output:

[2020-07-18 21:05:21] INFO: Aligning reads to the graph
0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100%
[2020-07-20 14:24:12] INFO: Aligned read sequence: 146639839881 / 277711296371 (0.52803)
[2020-07-20 14:24:12] INFO: Median overlap divergence: 0.0335728
[2020-07-20 14:26:07] INFO: Mean edge coverage: 19
[2020-07-20 14:26:35] INFO: Simplifying the graph

**Comment by mikolmogorov** [comment]
I see - this seems strange. Is it still writing the dot file? Could you post the output of `ls -l out_dir/20-repeat`? Potentially, there might be an issue with the graph traversing function...

**Comment by mikolmogorov** [comment]
@shanesturrock and also do `tail out_dir/20-repeat/graph_before_rr.gv`?

**Comment by shanesturrock** [comment]
The dot file was last written to on the 20th too:

`$ ls -l`
`total 1656`
`-rw-rw-r-- 1 shane shane 1691716 Jul 20 20:20 graph_before_rr.gv`

`$ tail graph_before_rr.gv`
`"35464" [style = "filled", fillcolor = "grey"];`
`"35465" [style = "filled", fillcolor = "grey"];`
`"35466" [style = "filled", fillcolor = "grey"];`
`"35467" [style = "filled", fillcolor = "grey"];`
`"35468" [style = "filled", fillcolor = "grey"];`
`"35469" [style = "filled", fillcolor = "grey"];`
`"35470" [style = "filled", fillcolor = "grey"];`
`"35471" [style = "filled", fillcolor = "grey"];`
`"35472" [style = "filled", fillcolor = "grey"];`
`"35473" [style = "filled", fillcolor = "grey"];`

RAM use is at 3.2TB now.

**Comment by mikolmogorov** [comment]
I see. Looks like it is indeed stuck in the function that generates the Graphviz output. Usually it takes seconds, but seems that it might be extremely inefficient for very large graphs.

I've created a version with graphviz output disabled - the code is in the `iss267` branch. It might makes sense to run it in parallel and see if it can catch up. The updated code should not use more than a few hundred Gb of RAM (theoretically).

**Comment by shanesturrock** [comment]
OK, I'll kill off the current job because it is getting to a massive amount of RAM now (pushing on for 4TB and growing at an alarming rate) and grab the modified branch so I can restart with that. Thanks for looking at this, I'll keep you posted on how it goes.

**Comment by shanesturrock** [comment]
I don't know if this helps but the previous build finally died with the following:

[2020-07-30 15:39:58] ERROR: Caught unhandled exception: std::bad_alloc
[2020-07-30 15:39:58] ERROR: 	flye-modules(_Z16exceptionHandlerv+0xd0) [0x449700]
[2020-07-30 15:39:58] ERROR: 	/lib64/libstdc++.so.6(+0x5e746) [0x7fee0c04d746]
[2020-07-30 15:39:58] ERROR: 	/lib64/libstdc++.so.6(+0x5e773) [0x7fee0c04d773]
[2020-07-30 15:39:58] ERROR: 	/lib64/libstdc++.so.6(+0x5e993) [0x7fee0c04d993]
[2020-07-30 15:39:58] ERROR: 	/lib64/libstdc++.so.6(_Znwm+0x7d) [0x7fee0c04df2d]
[2020-07-30 15:39:58] ERROR: 	flye-modules(_ZNSt6vectorIP9GraphEdgeSaIS1_EE19_M_emplace_back_auxIJRKS1_EEEvDpOT_+0x34) [0x4a9b14]
[2020-07-30 15:39:58] ERROR: 	flye-modules(_ZN15OutputGenerator9outputDotERKSt6vectorI15UnbranchingPathSaIS1_EERKSs+0xd3a) [0x4f850a]
[2020-07-30 15:39:58] ERROR: 	flye-modules(_Z11repeat_mainiPPc+0x1533) [0x501f73]
[2020-07-30 15:39:58] ERROR: 	flye-modules(main+0x59) [0x443b69]
[2020-07-30 15:39:58] ERROR: 	/lib64/libc.so.6(__libc_start_main+0xf5) [0x7fee0b50f555]
[2020-07-30 15:42:29] root: ERROR: Command '['flye-modules', 'repeat', '--disjointigs', '/raid/projects/active/nzgl01066/Flye-devel-assembly/Flye-2.6_Assembly-restart/flye_assembly/10-consensus/consensus.fasta', '--reads', 'mr.41.15.17.0.029.1.fa', '--out-dir', '/raid/projects/active/nzgl01066/Flye-devel-assembly/Flye-2.6_Assembly-restart/flye_assembly/20-repeat', '--config', '/raid/projects/active/nzgl01066/Flye-devel-assembly/Flye/flye/config/bin_cfg/asm_corrected_reads.cfg', '--log', '/raid/projects/active/nzgl01066/Flye-devel-assembly/Flye-2.6_Assembly-restart/flye_assembly/flye.log', '--threads', '128', '--min-ovlp', '2000', '--kmer', '17']' died with <Signals.SIGABRT: 6>.
[2020-07-30 15:42:29] root: ERROR: Pipeline aborted

I'm about to restart with the `iss267` branch

**Comment by shanesturrock** [comment]
I made a mistake when checking out the special branch (assumed the page I was look at it on would give me a download link to it but I actually just got the master again) so I've lost a week. Checked out the right version and we're now passed the point where it broke last time and it hasn't written the dot file. It 's now reached [SIMPL] Split nodes and has been sitting there for three days running on a single core.

**Comment by mikolmogorov** [comment]
@shanesturrock I see, so there might be more bottlenecks remaining.. If it not progressing, could you send me / post the latest log?

**Comment by shanesturrock** [comment]
It's currently using 1.4TB of RAM and 100% of a single CPU core. Here are the logs: https://we.tl/t-VOZsYSy8H1

Thanks again for this, very important that we get an assembly from this data and it is proving to be quite the challenge.

Shane

**Comment by shanesturrock** [comment]
It's moved on so it isn't stuck. Here's the tail end of the log now:

[2020-08-21 06:32:54] DEBUG:    Connection 1698173      -1689957        16     1
[2020-08-21 06:32:54] DEBUG:    Connection -1426768     -1426764        32     1
[2020-08-21 06:32:54] DEBUG:    Connection 754479       -1746757        8      1
[2020-08-21 06:32:54] DEBUG:    Connection -754211      1812954 3       1
[2020-08-21 06:32:54] DEBUG:    Connection 1118841      -646364 6       0.857143
[2020-08-21 06:32:54] DEBUG:    Connection 551858       -1190347        9      0.947368
[2020-08-21 06:32:54] DEBUG:    Connection -338 -337    10      1
[2020-08-21 06:38:05] DEBUG: [SIMPL] Resolved repeats: 216827
[2020-08-21 06:38:05] DEBUG: RR links: 4038352
[2020-08-21 06:38:05] DEBUG: Unresolved: 221

Still running single threaded, but still running so that's a plus!

**Comment by mikolmogorov** [comment]
Yay! It's a good sign, because now the pipeline will iterate a few simplification steps that you already passed - and every time they should be faseter computationally (at least, in theory).

Is it still using a lot of RAM?

Based on the logs, the current major slowdown is in the tip clipping function - I am planning to look into this in the near future.

**Comment by shanesturrock** [comment]
Cool. Still sitting on 1.4TB of RAM which isn't too bad for our machine given we have 6TB available. A few days on a single core isn't a disaster but more speed is still welcome!

**Comment by shanesturrock** [comment]
It's taken over three weeks now and it has iterated 9 times but SIMPL is still running. At iteration 9 I got this though:

[2020-09-10 16:45:42] DEBUG: [SIMPL] Resolved repeats: 0
[2020-09-10 16:45:42] DEBUG: RR links: 6054
[2020-09-10 16:45:42] DEBUG: Unresolved: 15
[2020-09-10 18:13:51] DEBUG: Removed 0 simple and 0 double chimeric junctions
[2020-09-10 18:20:00] DEBUG: [SIMPL] Collapsed 23300 haplotypes

The haplotypes message is new this time so I'm wondering how much more it is likely to do before it finishes the 'Simplifying the graph' step and moves on to the next step?

**Comment by mikolmogorov** [comment]
@shanesturrock that's promising! It means that the simplification iterations are over. There are a couple extra steps left, but they should be much faster than each iteration. So - fingers crossed - the repeat stage should finish soon.

**Comment by shanesturrock** [comment]
Another little step forward but it seems to be going slowly.

[2020-09-10 18:20:00] DEBUG: [SIMPL] Collapsed 23300 haplotypes
[2020-09-14 00:57:35] DEBUG:    Connection 1520075      -978212 1

That was about 8 hours ago and nothing more so it spent four days to write that.

**Comment by mikolmogorov** [comment]
Super slow indeed, but getting closer...

**Comment by shanesturrock** [comment]
Would you expect it to write many more of these? The previous iteration had this:

[2020-09-10 16:45:05] DEBUG: Total unique edges: 669748
[2020-09-10 16:45:42] DEBUG:    Connection 839639       800811  1       0.285714
[2020-09-10 16:45:42] DEBUG:    Connection -787855      825944  3       0.428571
[2020-09-10 16:45:42] DEBUG:    Connection 1787334      799158  1       0.5
[2020-09-10 16:45:42] DEBUG:    Connection 1015377      43307   1       0.5
[2020-09-10 16:45:42] DEBUG:    Connection 697723       1220999 2       0.5
[2020-09-10 16:45:42] DEBUG:    Connection -1306174     968825  3       0.2
[2020-09-10 16:45:42] DEBUG:    Connection -1335750     1060515 2       0.444444
[2020-09-10 16:45:42] DEBUG:    Connection -780788      -257524 2       0.444444
[2020-09-10 16:45:42] DEBUG:    Connection 103393       1019098 3       0.5
[2020-09-10 16:45:42] DEBUG:    Connection -989847      -485123 2       0.4
[2020-09-10 16:45:42] DEBUG:    Connection 922825       1147223 3       0.5
[2020-09-10 16:45:42] DEBUG:    Connection -989260      -611808 3       0.352941
[2020-09-10 16:45:42] DEBUG:    Connection -1148035     -536241 2       0.285714
[2020-09-10 16:45:42] DEBUG:    Connection -1050063     944480  1       0.5
[2020-09-10 16:45:42] DEBUG:    Connection -461201      -392922 1       0.5
[2020-09-10 16:45:42] DEBUG: [SIMPL] Resolved repeats: 0
[2020-09-10 16:45:42] DEBUG: RR links: 6054
[2020-09-10 16:45:42] DEBUG: Unresolved: 15
[2020-09-10 18:13:51] DEBUG: Removed 0 simple and 0 double chimeric junctions

Hoping it isn't going to write that many connections taking four days each.....

**Comment by mikolmogorov** [comment]
No, there are two a bit different repeat resolution algorithms - so connections should be different. Not sure why it's taking so long though..

**Comment by shanesturrock** [comment]
Huge genome with lots of repeats. I'll let it run and will update you as it does more.

**Comment by mikolmogorov** [comment]
Thanks! Do you mind uploading the latest log again? I'll try to fix the bottleneck in the graph simplification iterations in case you need to restart the stage (hopefully not though).

**Comment by shanesturrock** [comment]
Here you go: https://wsi.li/dl/NjiTrnQmgHCXFna3G/

**Comment by shanesturrock** [comment]
Two more:

[2020-09-15 19:10:02] DEBUG:    Connection -834547      61354   77
[2020-09-15 19:10:02] DEBUG:    Connection -1099575     1099576 7

Still running.

**Comment by mikolmogorov** [comment]
@shanesturrock thanks for sending the log files!

Looks like there is an unexpected slowdown in the tip clipping function. E.g., Flye took about a week to build the graph and align reads, but the reset of ~3.5 weeks of graph simplifications were spent inside the tip clipping procedure!

I've just made some optimizations, and I expect the graph simplification to proceed much faster for you now. E.g., the graph simplification should *theoretically* run within a day with this update (instead of weeks). I've put the updates to the `iss267` branch that you are using, and also disabled the extra repeat resolution function that is currently causing troubles.

If you have resources, I would recommend to run the updated version in parallel. It should take about a week to reach the graph simplification phase, but then go much faster.

**Comment by shanesturrock** [comment]
Thanks for doing this, really appreciated. I'm making a duplicate of the current working directory and I'll run this latest version in there. Plenty of CPU and RAM to spare so it shouldn't be a problem.

**Comment by shanesturrock** [comment]
OK, so both runs have progressed. The updated version is working through the SIMPL steps again but the previous version has written the contigs.fasta file which is exciting. However, when I check using bbtools stats.sh, there only appears to be about 8Gbp total contig length down from 22Gbp prior to the repeat resolution and the genome is expected to be around 25Gbp. It looks like the assembly has collapsed due to the repeats. Given that the input data came from MaSuRCA and we have already discussed running the job again using the raw pacbio I wanted to get your advice on this and if helpful I can again provide the logs. I'll let the jobs continue for now.

**Comment by shanesturrock** [comment]
Another update and a question. I've looked at the stats for the consensus.fasta file in 10-consensus and that's showing around 22Gbp but the graph_before_rr.fasta in 20-repeat is only 9Gbp and repeat_graph_edges.fasta is 22Gbp. In 30-contigger, the contigs.fasta is 8Gbp and it has written the graph_final.gv and scaffolds_links.txt but is now stuck writing the dot file (I assume this is the gfa where it is somewhat expected to fail.)

So, I was hoping that the 9Gbp that went into the repeat resolution and then into the contigs wasn't all the sequence that will end up in the final assembly and the missing sequence will be pulled from the consensus and that if it is stuck at the gfa writing stage I can kill it and restart from the polishing step? Somewhat stressed that the assembly appears to be lacking so much of the expected genome sequence after the consensus looked so promising.

**Comment by mikolmogorov** [comment]
@shanesturrock first of all, I'm glad that at least one assembly is now finished :) How is the most recent one going - is it faster than the one that just finished, or not so much?

Yes, I expect that generating the final gfa may take forever, so it makes sense to kill the process and polish the generated contigs manually.

Wrt to the size - it is indeed concerning. One way to check could be to run BUSCO - if it just the collapsed repeats, it should find high number of core genes.

I think that size reduction is likely related to the fact that only ~50% reads were aligned to the graph because of wrong thresholds. I've now updated iss267 branch and changed the threshold from 3% to 15% - that should allow most of the reads to align. I did not make this change before because I am not sure how it may affect the overall speed, and wanted at least the first assembly iteration to finish.

I think it makes sense to make another run with the updated config. In this case, you might want to double check that the proper updated config is loaded by grepping the log file for `read_align_ovlp_divergence` string: the updated version should show value of 0.15, instead of 0.03.

**Comment by shanesturrock** [comment]
Hi Mikhail,

Unfortunately, we just had a hardware failure on our RAID so the jobs that were running have stopped. From what I could see of the new version, it wasn't going through the SIMPL stages any quicker as it had taken most of the week to complete the first iteration. I'll restart the completed contigs run from the polishing step.

I've used BUSCO previously with other assemblies done with SOAPdenovo2 from Illumina data so it makes sense to try here too so I'll do that when the machine is back up and running and I've checked out the new version. All the data is on the RAID and inaccessible at the moment.

One observation I've had from previous assembly attempts with the PacBio reads was that if the overlap assembly step (tested using CANU) used too short an overlap I got quite obvious repeat collapse tested by mapping 10kbp insert mate pair illumina reads. Mapping those against the raw PacBio reads shows exactly the expected single insert size but with overlap assembly using 1Kbp I could see multiple insert size peaks and it was only when the overlaps were 5Kbp did it revert to a single peak at 10k. This suggests some very long runs of repeated sequence which was one of the primary reasons for going with the PacBio reads. We've estimated around 80+% repeats. Repeats within a long read shouldn't be an issue and shouldn't be broken up I would think.

**Comment by shanesturrock** [comment]
Hardware issues appear to be resolved but we're letting the system run for a while before calling it done. I've checked out the latest code and the value has changed to 0.15 and I'll restart that job shortly. I'll also start the polishing step on the other assembly.

**Comment by shanesturrock** [comment]
Sorry for not getting back earlier. We're still having hardware issues so I haven't been able to continue with the jobs. However, I did run BUSCO on the contigs as you suggested. Using the 303 eukaryotes set, it gets C:141 (S:111, D:30), F:71, M:91, n:303. It would make sense if more than half the genome is missing for this to be the case, but I've also run BUSCO on the Pinus taeda V2.0 assembly which is their most complete and get very similar figures from BUSCO (C:159 [S:107, D:52], F:51, M:93) so it may be an issue with BUSCO and these large genomes rather than the data itself. We have a transcriptome set and that gets 299 out of 303 complete with a couple of fragments and two missing entirely. Meanwhile, I'm doing what I can on the hardware while we wait for a permanent fix and then I'll get the assembly proper running again.

**Comment by shanesturrock** [comment]
Despite the hardware issues, the system has kept running long enough to get to the simplifying the graph step and we've got a value for aligned read sequence:

[2020-10-21 13:45:03] INFO: Aligned read sequence: 274696524376 / 277711296371 (0.989144)
[2020-10-21 13:45:03] INFO: Median overlap divergence: 0.0335724

This is what we got last time:

[2020-08-16 13:52:16] INFO: Aligned read sequence: 146639839881 / 277711296371 (0.52803)
[2020-08-16 13:52:16] INFO: Median overlap divergence: 0.0335728

So, we've gone from 52% of reads aligned to 99%. Hopefully this means we'll be able to get more than just the 8Gbp worth of contigs out but I'm expecting the machine will need another restart before we get there since it took six weeks to complete the graph simplification step and get into the contigger stage last time. A set of contigs with the full 22+Gbp would be awesome.

**Comment by shanesturrock** [comment]
I've checked the stats on the graph_before_rr.fasta file for the old and new runs and I was expecting the new one to have more for the total contig length but they're both about 9Gbp even though the new run has used all the reads. I'm assuming this means the final assembly is still going to be significantly lacking length like the last one was even though it has used nearly all the reads this time. I would like to be wrong about this and if I am it would be good for some reassurance. The alternative is to stop trying to restart the assembly and go back to the raw PacBio (100 million reads) and assemble them from the start without error correction but I'm concerned how long such an assembly is likely to take before I embark on it.

**Comment by mikolmogorov** [comment]
@shanesturrock thanks for the update!

You are right, the size of the graph before repeat resolution is not expected to change. But think about it as a "genome" with all repeats collapsed. Since your genome has such high repeat content, a collapsed representation could be much smaller that the actual genome. My hope is that with more aligned reads, many more repeats should be resolved, which should increase the assembly size. I think it does make sense to let it finish. 

On the other hand, restarting from raw PacBio makes sense as well, if you have resources to run it in parallel. Have you tried other fast assemblers, such as wtdbg2?

Wrt to BUSCO analysis - I suspect that some genes might be missing because of indel errors. Since you have ~15x PB coverage, I would expect ~1% consensus error, which is substantial. Do you have Illumina data that you can use for polishing? If you have a transcript set, you might have better luck aligning the gene sequences against the assembly with any general purpose aligner.

**Comment by shanesturrock** [comment]
We may be able to run both a fresh assembly with raw PacBio and keep this one running in parallel if and when we get the all clear on the hardware. At the moment we're still expecting to have to do another restart so I'm just running this to see what it gets and not expecting completion. It would be good if it could expand this assembly and I'm willing to let it run through. There was the issue someone else saw where the process got stuck in the contig generation step and needed to be restarted at polishing but when I tried to restart like that it told me that there was no polishing stage to start and the best I could do was get it to restart the contigger step.

We have around 25x PacBio and another 80x of Illumina which is what was used to error correct. I could certainly use that for polishing after the fact. We've been focussed initially on MaSuRCA but moved to FLYE on Aleski's recommendation. I'm certainly open to trying others of course. I had tried MARVEL but that barfed at the amount of repeats we had, and the usual suspects like CANU just collapsed the repeats down. If it makes sense to try it I'll give wtdbg2 a go given it was able to handle Axolotl which is even a bit bigger than our genome.

I'm more surprised that BUSCO got similar results with the v2 Pinus taeda genome which is already published to be honest and that's what makes me think it has a problem with these large genomes. With the transcript set we have though I could likely do some comparisons between that and the genome sequence. Compute resources are a bit strained and each thing I try takes up a big amount of time and we only have the one (admittedly big) machine.

**Comment by mikolmogorov** [comment]
@shanesturrock ah, I completely forgot that the reads are error-corrected with MaSuRCA - so that the read base error should be low. We've seen BUSCO reporting many missing genes for seemingly complete assemblies of "unusual" genomes - so it could be that  your genome is one of these. Btw maybe it makes sense to run BUSCO on transcripts set for the sake of comparison?

There is also a new read error-corrector called Ratatosk https://github.com/DecodeGenetics/Ratatosk. It seems to be doing a good job on human genomes, maybe it makes sense to give it a try.

**Comment by shanesturrock** [comment]
@fenderglass From what we've seen in previous runs of FLYE, the error rate in the data that came out of MaSuRCA is still quite high with some reads seemingly not corrected at all. I've also used FMLRC to error correct them using the Illumina reads and that didn't do any better as we were still seeing a lot of uncorrected reads not to mention it took an enormous amount of CPU time. That's why I'm leaning towards using uncorrected reads and then polishing after assembly. With 25x PacBio it should be possible to get a decent assembly even with 10% error rates. I think once we're over the hardware issues I'll give wtdbg2 a go as the performance promises to be pretty good.

BUSCO produces an almost perfect result with our transcriptome assembly so I tend to agree that it's struggling with the assembled genomes for some reason.

**Comment by shanesturrock** [comment]
The new version of FLYE you did got through the graph simplification step way quicker than previously doing it in just 6 days. It's generated contigs but using stats.sh on that shows that there's still only 12Gbp of genome represented in the contigs.fasta using bbtools stats.sh to get the stats) but on the plus side there are more contigs and there's 122 over 500Kbp long. Worth running from scratch with raw PacBio but I'm testing with wtdbg2 as you suggested which seems to be very fast with the various genome sets on our hardware and performance is in line with what they have in their table of examples. I'll try with our full data set and see if it gets an answer in about a week like they got with Axolotl. Still keen to keep FLYE going though.

**Comment by mikolmogorov** [comment]
@shanesturrock glad to hear! 12 Gb is still way less than expected, but increase from 9 Gb is substantial too, plus indeed there should be a bump in contiguity since many more repeats are resolved.

Definitely worth trying with raw PacBio reads. I remember you mentioned that previous Flye versions failed on this one, but hopefully the recent performance optimization will help.

**Comment by shanesturrock** [comment]
I'm having some decent success with wtdbg2, it has been able to do an assembly of the 20x PacBio data in just 3 days. I'm now giving it the full set but it looks like speed isn't a problem although I don't know for sure how good the assembly is until I run BUSCO on it.

Meanwhile, FLYE reached the gfa writing stage in 30-contigger and has been writing that file for several days so I've killed it before it fills up too much storage as it was 605GB already. I still plan on going back to the raw reads with FLYE too of course but for now I'll focus on wtdbg2. Feel free to close this case because it has got pretty long although a fix for the massive gfa file would be welcome. Perhaps a flag that toggles writing that off.

**Comment by mikolmogorov** [comment]
Sounds good! Let me know if you have more questions. Gfa generation should now be fixed in the development branch (https://github.com/fenderglass/Flye/issues/290) - so hopefully not going to be a problem in the future.
